File: tests/python/public_api.py
"""Python 公共 API 跨层契约测试。

覆盖 Prepared TQL、一等 Path/List、四类属性索引、存储诊断和历史入口迁移错误。
"""

import json
import os
import tempfile

import triviumdb


def run_shared_contract(root: str) -> None:
    contract_path = os.path.join(os.path.dirname(__file__), "..", "contracts", "public_cases.json")
    with open(contract_path, encoding="utf-8") as source:
        contract = json.load(source)
    assert contract["schema_version"] == 1
    path = os.path.join(root, "shared-contract.tdb")
    db = triviumdb.TriviumDB(path, dim=contract["setup"]["dimension"])
    for node in contract["setup"]["nodes"]:
        db.insert_with_id(node["id"], node["vector"], node["payload"])
    for edge in contract["setup"]["edges"]:
        db.link(edge["source"], edge["target"], edge["label"], edge["weight"])
    for case in contract["cases"]:
        expected = case["expected"]
        if case["operation"] == "get_payload":
            payload = db.get_payload(case["node_id"])
            assert payload[expected["field"]] == expected["value"], case["name"]
        elif case["operation"] == "prepared_tql":
            prepared = db.prepare_tql(case["tql"])
            rows = db.execute_prepared_tql(prepared, case.get("parameters", {}))
            assert len(rows) == expected["row_count"], case["name"]
            assert rows[0].row[expected["column"]] == expected["value"], case["name"]
        elif case["operation"] == "tql_path":
            rows = db.tql(case["tql"])
            assert rows[0].row["path"] == expected["path"], case["name"]
        elif case["operation"] == "prepared_missing_parameter":
            prepared = db.prepare_tql(case["tql"])
            try:
                db.execute_prepared_tql(prepared, case.get("parameters", {}))
            except RuntimeError as error:
                assert any(needle in str(error) for needle in expected["error_contains_any"])
            else:
                raise AssertionError(case["name"])
        else:
            raise AssertionError(f"未知共享契约操作: {case['operation']}")
    db.close()


def main() -> None:
    with tempfile.TemporaryDirectory(prefix="triviumdb-python-public-api-") as root:
        run_shared_contract(root)
        path = os.path.join(root, "api.tdb")
        db = triviumdb.TriviumDB(path, dim=2)
        db.insert_with_id(1, [1.0, 0.0], {"kind": "a", "score": 1, "tenant": "t", "region": "r"})
        db.insert_with_id(2, [0.9, 0.1], {"kind": "b", "score": 2, "tenant": "t", "region": "r"})
        db.link(1, 2, "next", 1.0)

        stages: list[str] = []

        class FullHook:
            def on_pre_search(self, query, ctx):
                stages.append("pre")
                return query

            def on_custom_recall(self, query, ctx):
                stages.append("custom")
                return None

            def on_post_recall(self, hits, ctx):
                stages.append("post_recall")
                return hits

            def on_pre_graph_expand(self, hits, ctx):
                stages.append("pre_graph")
                return hits

            def on_rerank(self, hits, ctx):
                stages.append("rerank")
                return hits

            def on_post_search(self, hits, ctx):
                stages.append("post")
                return hits

        db.set_hook(FullHook())
        db.search_with_context([1.0, 0.0], top_k=2, expand_depth=1, min_score=0.0)
        assert {"pre", "custom", "post_recall", "pre_graph", "rerank", "post"} <= set(stages)
        db.clear_hook()

        class BrokenHook:
            def on_pre_search(self, query, ctx):
                raise RuntimeError("python-hook-failure")

        db.set_hook(BrokenHook())
        try:
            db.search_with_context([1.0, 0.0], top_k=2)
            raise AssertionError("Hook 异常必须传播")
        except RuntimeError as error:
            assert "python-hook-failure" in str(error)
        db.clear_hook()

        db.create_index("kind")
        db.create_ordered_index("score")
        db.create_composite_index(["tenant", "region"])
        db.create_bitmap_index("region")
        indexes = db.index_info()
        assert len(indexes) == 4
        assert next(item for item in indexes if item["kind"] == "composite")["fields"] == [
            "tenant",
            "region",
        ]

        storage = db.storage_info()
        assert storage["database_format_current"] == 9
        assert storage["property_index_format"] == 6
        assert storage["node_count"] == 2

        vector_prepared = db.prepare_tql(
            "SEARCH VECTOR [$x, $y] TOP 1 RETURN *"
        )
        assert vector_prepared.parameter_names() == ["x", "y"]
        vector_rows = db.execute_prepared_tql(vector_prepared, {"x": 1, "y": 0.0})
        assert vector_rows[0].row["_"].get("id") == 1

        prepared = db.prepare_tql(
            'FIND {kind: "a"} AS seed WITH seed RETURN seed, $bonus + 1 AS score'
        )
        assert prepared.parameter_names() == ["bonus"]
        prepared_rows = db.execute_prepared_tql(prepared, {"bonus": 4})
        assert len(prepared_rows) == 1
        assert prepared_rows[0].row["score"] == 5
        try:
            db.execute_prepared_tql(prepared, {})
        except RuntimeError as error:
            assert "缺少参数" in str(error) or "missing parameter" in str(error)
        else:
            raise AssertionError("缺少 Prepared TQL 参数必须失败")

        rows = db.tql(
            "SEARCH VECTOR [1, 0] TOP 1 AS seed WITH seed shortest_paths seed TO [2] "
            "LABEL next AS route WITH route RETURN path(route) AS path"
        )
        assert rows[0].row["path"] == [1, 2]

        try:
            db.tql_mut('FIND {kind: "a"} RETURN *')
        except RuntimeError as error:
            assert "API 已移除" in str(error) or "API removed" in str(error)
        else:
            raise AssertionError("tql_mut 读查询必须失败")
        db.close()


if __name__ == "__main__":
    main()

File: tests/node/public_api.js
'use strict'

// Node 公共 API 跨层契约测试：覆盖 Prepared TQL、索引/存储诊断、
// u64 字符串映射、Payload Filter 和已移除入口的稳定迁移错误。

const assert = require('node:assert/strict')
const fs = require('node:fs')
const os = require('node:os')
const path = require('node:path')
const contract = require('../contracts/public_cases.json')
const { TriviumDB } = require('../../index.js')

const root = fs.mkdtempSync(path.join(os.tmpdir(), 'triviumdb-node-public-api-'))
const file = path.join(root, 'api.tdb')

function runSharedContract(root) {
  assert.equal(contract.schema_version, 1)
  const db = new TriviumDB(path.join(root, 'shared-contract.tdb'), {
    dim: contract.setup.dimension,
  })
  for (const node of contract.setup.nodes) {
    db.insertWithId(node.id, node.vector, node.payload)
  }
  for (const edge of contract.setup.edges) {
    db.link(edge.source, edge.target, edge.label, edge.weight)
  }
  for (const testCase of contract.cases) {
    const expected = testCase.expected
    if (testCase.operation === 'get_payload') {
      const payload = db.getPayload(testCase.node_id)
      assert.equal(payload[expected.field], expected.value, testCase.name)
    } else if (testCase.operation === 'prepared_tql') {
      const prepared = db.prepareTql(testCase.tql)
      const rows = db.executePreparedTql(prepared, testCase.parameters || {})
      assert.equal(rows.length, expected.row_count, testCase.name)
      assert.equal(rows[0][expected.column], expected.value, testCase.name)
    } else if (testCase.operation === 'tql_path') {
      const rows = db.tql(testCase.tql)
      assert.deepEqual(rows[0].path, expected.path.map(String), testCase.name)
    } else if (testCase.operation === 'prepared_missing_parameter') {
      const prepared = db.prepareTql(testCase.tql)
      assert.throws(
        () => db.executePreparedTql(prepared, testCase.parameters || {}),
        error => expected.error_contains_any.some(needle => String(error).includes(needle)),
        testCase.name,
      )
    } else {
      throw new Error(`未知共享契约操作: ${testCase.operation}`)
    }
  }
  db.close()
}

try {
  runSharedContract(root)
  assert.throws(
    () => new TriviumDB(file, 2),
    /object|对象|options/i,
    '旧位置参数构造器必须明确拒绝',
  )

  const db = new TriviumDB(file, { dim: 2 })
  db.insertWithId(1, [1, 0], { kind: 'a', score: 1, tenant: 't', region: 'r' })
  db.insertWithId(2, [0.9, 0.1], { kind: 'b', score: 2, tenant: 't', region: 'r' })
  db.link(1, 2, 'next', 1)

  const stages = []
  db.setHook({
    onPreSearch(query) {
      stages.push('pre')
      return query
    },
    onCustomRecall() {
      stages.push('custom')
      return null
    },
    onPostRecall(hits) {
      stages.push('postRecall')
      return hits
    },
    onPreGraphExpand(hits) {
      stages.push('preGraph')
      return hits
    },
    onRerank(hits) {
      stages.push('rerank')
      return hits
    },
    onPostSearch(hits) {
      stages.push('post')
      return hits.slice(0, 1)
    },
  })
  const hooked = db.searchWithContext([1, 0], {
    topK: 2,
    expandDepth: 1,
    minScore: 0,
  })
  assert.equal(hooked.hits.length, 1)
  assert.deepEqual(new Set(stages), new Set([
    'pre', 'custom', 'postRecall', 'preGraph', 'rerank', 'post',
  ]))
  db.clearHook()

  db.setHook({
    onPreSearch() {
      throw new Error('node-hook-failure')
    },
  })
  assert.throws(
    () => db.searchWithContext([1, 0], { topK: 2 }),
    /TDB_HOOK_EXECUTION|node-hook-failure/,
  )
  db.clearHook()

  db.createIndex('kind')
  db.createOrderedIndex('score')
  db.createCompositeIndex(['tenant', 'region'])
  db.createBitmapIndex('region')
  const indexes = db.indexInfo()
  assert.equal(indexes.length, 4)
  assert.deepEqual(indexes.find(item => item.kind === 'composite').fields, ['tenant', 'region'])

  const storage = db.storageInfo()
  assert.equal(storage.database_format_current, 9)
  assert.equal(storage.property_index_format, 6)
  assert.equal(storage.node_count, 2)

  const vectorPrepared = db.prepareTql('SEARCH VECTOR [$x, $y] TOP 1 RETURN *')
  assert.deepEqual(vectorPrepared.parameterNames(), ['x', 'y'])
  const vectorRows = db.executePreparedTql(vectorPrepared, { x: 1, y: 0 })
  assert.equal(vectorRows[0]._.id, '1')

  const prepared = db.prepareTql('FIND {kind: "a"} AS seed WITH seed RETURN seed, $bonus + 1 AS score')
  assert.deepEqual(prepared.parameterNames(), ['bonus'])
  const preparedRows = db.executePreparedTql(prepared, { bonus: 4 })
  assert.equal(preparedRows.length, 1)
  assert.equal(preparedRows[0].score, 5)
  assert.throws(() => db.executePreparedTql(prepared, {}), /missing parameter|缺少参数/)

  const pathRows = db.tql('SEARCH VECTOR [1, 0] TOP 1 AS seed WITH seed shortest_paths seed TO [2] LABEL next AS route WITH route RETURN path(route) AS path')
  assert.deepEqual(pathRows[0].path, ['1', '2'])

  const advancedFiltered = db.searchAdvanced([1, 0], {
    topK: 2,
    minScore: 0,
    payloadFilter: { kind: 'b' },
  })
  assert.equal(advancedFiltered.length, 1)
  assert.equal(advancedFiltered[0].payload.kind, 'b')

  const contextFiltered = db.searchWithContext([1, 0], {
    topK: 2,
    minScore: 0,
    payloadFilter: { kind: 'b' },
  })
  assert.equal(contextFiltered.hits.length, 1)
  assert.equal(contextFiltered.hits[0].payload.kind, 'b')

  const basicFiltered = db.search([1, 0], 2, 0, 0, { kind: 'b' })
  assert.equal(basicFiltered.length, 1)
  assert.equal(basicFiltered[0].payload.kind, 'b')

  assert.throws(
    () => db.searchAdvanced([1, 0], { payloadFilter: { score: { $gt: true } } }),
    /payloadFilter|过滤器|Filter/i,
  )

  assert.throws(
    () => db.tqlMut('FIND {kind: "a"} RETURN *'),
    /TDB_API_MIGRATION_REQUIRED/,
  )
  db.close()
} finally {
  fs.rmSync(root, { recursive: true, force: true })
}
